utils/land-patterns.stanza

```diff
@@ -239,52 +298,123 @@ public pcb-landpattern non-plated-hole-pkg (drill-r:Double, mask-r:Double) :
-  val w2 = w / 2.0
-  val h2 = h / 2.0
-  Line(line-w, [
-    Point(xc - w2, yc - h2)
-    Point(xc + w2, yc - h2)
-    Point(xc + w2, yc + h2)
-    Point(xc - w2, yc + h2)
-    Point(xc - w2, yc - h2)])
-
-public defn LineRectangle (w:Double, h:Double, xc:Double, yc:Double) :
-  LineRectangle(w, h, xc, yc, 0.254)
 
-public defn LineRectangle (w:Double, h:Double) :
-  LineRectangle(w, h, 0.0, 0.0, 0.254)
+;===============================================================================
+;===================== Generic Landpattern Definition ==========================
+;===============================================================================
+; PinPadMapping is a not-yet-placed pad, with assigned name, pad, and pose
+public pcb-struct ocdb/land-patterns/PinPadMapping:
+  pad-name:  Ref,  ; the name of the pad
+  pad-shape: JITXDef, ; the corresponding pcb-pad
+  pose:      Pose,    ; the pose of the pad
 
-public defn PlusSymbol (x:Double, y:Double) :
-  Union([
-    Line(0.254, [ Point(x - 0.5, y), Point(x + 0.5, y) ])
-    Line(0.254, [ Point(x, y - 0.5), Point(x, y + 0.5) ])
-  ])
-; ====== Pad Collection Helpers ================================================
 
-; 2D array of pins
-public defn make-pin-grid (n-pin:Int, n-row:Int, pin-pitch:Double, row-pitch:Double, pad-ref) :
+; A generic landpattern created from a sequence of PinPadMappings
+public defn make-generic-landpattern (
+  pads:Tuple<PinPadMapping>
+):
   inside pcb-landpattern:
-    
-    val n-col = to-int(ceil(to-double(n-pin) / to-double(n-row)))
-
-    val l0 = loc((to-double(n-row - 1) * row-pitch) / -2.0, (to-double(n-col - 1) * pin-pitch) / 2.0)
-    
-    for i in 0 to n-col do :
-      for j in 0 to n-row do :
-        pad p[n-row * i + j + 1] : pad-ref at l0 * loc(to-double(j) * row-pitch, (- to-double(i) * pin-pitch))
-    
-public defn make-pin-grid (n-pin:Int, n-row:Int, pin-pitch:Double, row-pitch:Double) :
-  make-pin-grid(n-pin, n-row, pin-pitch, row-pitch, pth-pad(0.5))
+    for p in pads do:
+      pad (pad-name(p)): (pad-shape(p) as Pad) at pose(p)
+
+; Create a generic landpattern with a uniform pad
+public defn make-generic-landpattern (
+  pad-names:Seqable<Ref>,
+  poses:Seqable<Pose>,
+  pad-shape:JITXDef
+):
+  val pads = 
+    for (pn in pad-names, pp in poses) seq:
+      PinPadMapping(pn, pad-shape, pp) 
+  make-generic-landpattern(to-tuple(pads))
+
+;===============================================================================
+;========================= Pad Arrangements ====================================
+;===============================================================================
+; Naming for a 2-column arrangment of pins 
+public defn soic-naming-convention (row:Int, column:Int, num-pins:Int) -> Ref:
+  if column == 0:
+    IndexRef(Ref("p"), row + 1)
+  else if column == 1:
+    IndexRef(Ref("p"), num-pins - row)
+  else:
+    fatal("soic naming convention supports a maximum of 2 columns")
```

Comment:
I would put it first. It is a precondition for this function

Reply:
done